src/main/java/example/TriangleCount.java

```diff
@@ -0,0 +1,210 @@
+package example;
+
+import org.apache.flink.api.common.ProgramDescription;
+import org.apache.flink.api.common.functions.FlatMapFunction;
+import org.apache.flink.api.common.functions.MapFunction;
+import org.apache.flink.api.common.functions.ReduceFunction;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.api.java.tuple.Tuple1;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.graph.Edge;
+import org.apache.flink.graph.EdgeDirection;
+import org.apache.flink.graph.Graph;
+import org.apache.flink.graph.NeighborsFunctionWithVertexValue;
+import org.apache.flink.graph.Triplet;
+import org.apache.flink.graph.Vertex;
+import org.apache.flink.types.NullValue;
+import org.apache.flink.util.Collector;
+import util.TriangleCountData;
+
+import java.util.HashSet;
+import java.util.Iterator;
+
+public class TriangleCount implements ProgramDescription {
+
+	public static void main(String [] args) throws Exception {
+
+		if(!parseParameters(args)) {
+			return;
+		}
+
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+
+		DataSet<Edge<String, NullValue>> edges = getEdgesDataSet(env);
+
+		Graph<String, NullValue, NullValue> graph = Graph.fromDataSet(edges, env);
+
+		// simulate the first superstep
+		// select the neighbors with id greater than the current vertex's id
+		DataSet<Vertex<String, HashSet<String>>> verticesWithLowerNeighbors =
+				graph.groupReduceOnNeighbors(new GatherLowerNeighbors(), EdgeDirection.ALL);
+
+		Graph<String, HashSet<String>, NullValue> graphWithVertexNeighbors = Graph.fromDataSet(verticesWithLowerNeighbors,
+				edges, env);
+
+		// simulate the second superstep
+		// propagate each received "message" to neighbours with higher id
+		DataSet<Vertex<String, HashSet<String>>> verticesWithPropagatedValues =
+				graphWithVertexNeighbors.groupReduceOnNeighbors(new PropagateNeighborValues(), EdgeDirection.ALL);
+
+		Graph<String, HashSet<String>, NullValue> graphWithPropagatedVertexValues = Graph.fromDataSet(verticesWithPropagatedValues,
+				edges, env);
+
+		DataSet<Vertex<String, Integer>> verticesWithNumberOfTriangles = graphWithPropagatedVertexValues.getTriplets()
```

Comment:
For the same reasons, we should avoid `getTriplets()`, which spreads the computation on the edges. Remember, we want to write this is a "vertex-centric" way :)
Each vertex has to check for triangles by comparing its msgs and its neighbors.